File: metasystem/internal/ui/project/links.go
package project

import (
	"net/url"
	"path"
	"strings"

	"github.com/widoriezebos/agentic-tools/metasystem/internal/ui/markdown"
)

// classify raises the relative Markdown links of a parsed tree to document
// references. Every other href keeps the classification the parser gave it, so
// a scheme this interface does not open stays unresolved and is rendered as
// text rather than as an anchor.
func classify(blocks []markdown.Block, from string) {
	for index := range blocks {
		block := &blocks[index]
		classifyInlines(block.Inlines, from)
		classify(block.Blocks, from)
		for item := range block.Items {
			classify(block.Items[item].Blocks, from)
		}
		for cell := range block.Head {
			classifyInlines(block.Head[cell], from)
		}
		for row := range block.Rows {
			for cell := range block.Rows[row] {
				classifyInlines(block.Rows[row][cell], from)
			}
		}
	}
}

func classifyInlines(inlines []markdown.Inline, from string) {
	for index := range inlines {
		inline := &inlines[index]
		classifyInlines(inline.Inlines, from)
		if inline.Type != "link" || inline.Target != markdown.TargetUnresolved {
			continue
		}
		if id, ok := DocumentID(from, inline.Href); ok {
			inline.Target = markdown.TargetDocument
			inline.ID = id
		}
	}
}

// DocumentID resolves one href against the directory of the document it was
// written in, and reports the id of the document it names.
//
// It answers false for anything this interface does not serve as a document:
// an absolute path, a scheme, a file that is not Markdown, a name that climbs
// out of the checkout, and a name beneath a directory the route refuses. The
// admissibility test is the route's own, so a link is offered only when the
// route would answer it, and a link the route would refuse is text.
func DocumentID(from, href string) (string, bool) {
	target, _, _ := strings.Cut(href, "#")
	target, _, _ = strings.Cut(target, "?")
	if target == "" || strings.HasPrefix(target, "/") || hasScheme(target) {
		return "", false
	}
	if decoded, err := url.PathUnescape(target); err == nil {
		target = decoded
	}
	resolved := path.Join(path.Dir(from), target)
	if !admissibleID(resolved) {
		return "", false
	}
	return resolved, true
}

// hasScheme reports whether a href begins with a URL scheme, which is the one
// shape a relative path can never have: a colon before the first slash.
func hasScheme(href string) bool {
	for index, character := range href {
		switch {
		case character == ':':
			return index > 0
		case character == '/' || character == '.' || character == '#' || character == '?':
			return false
		case isSchemeCharacter(character, index):
			continue
		default:
			return false
		}
	}
	return false
}

func isSchemeCharacter(character rune, index int) bool {
	letter := (character >= 'a' && character <= 'z') || (character >= 'A' && character <= 'Z')
	if index == 0 {
		return letter
	}
	digit := character >= '0' && character <= '9'
	return letter || digit || character == '+' || character == '-'
}

File: metasystem/internal/ui/web/_app/src/project/links.ts
/**
 * The belt.
 *
 * The engine already classified every link, and the policy already refuses an
 * inline script and a javascript: navigation. This is the third wall, and it
 * is here because it is the cheapest of the three: a link becomes an anchor
 * only when this file agrees, and everything else is rendered as text.
 *
 * The rules are the reader's own, written again: an external link is only the
 * two spellings a browser may open, and a document link resolves against the
 * directory of the document it was written in, stays beneath the checkout, is
 * Markdown, and names no directory the route refuses.
 */

/** The directories the route never serves, whatever the letter case. */
const REFUSED = [".git", "node_modules", "artifacts", "bin"];

/** The href a new tab may be opened with, or null for anything else. */
export function externalHref(href: string): string | null {
  return href.startsWith("http://") || href.startsWith("https://") ? href : null;
}

/** The heading an in-page link points at, or null. */
export function fragmentOf(href: string): string | null {
  return href.startsWith("#") && href.length > 1 ? href.slice(1) : null;
}

/**
 * The id of the document a relative href names, or null when this build would
 * not serve it.
 */
export function documentIdFor(from: string, href: string): string | null {
  const target = href.split("#")[0].split("?")[0];
  if (target === "" || target.startsWith("/") || hasScheme(target)) {
    return null;
  }
  const resolved = resolve(directoryOf(from), decode(target));
  return resolved !== null && admissible(resolved) ? resolved : null;
}

function decode(target: string): string {
  try {
    return decodeURIComponent(target);
  } catch {
    return target;
  }
}

function directoryOf(id: string): string[] {
  const segments = id.split("/");
  segments.pop();
  return segments;
}

/** Joins a relative path onto a directory, or null when it climbs out. */
function resolve(directory: string[], target: string): string | null {
  const segments = [...directory];
  for (const segment of target.split("/")) {
    if (segment === "" || segment === ".") {
      continue;
    }
    if (segment === "..") {
      if (segments.length === 0) {
        return null;
      }
      segments.pop();
      continue;
    }
    segments.push(segment);
  }
  return segments.length === 0 ? null : segments.join("/");
}

function admissible(id: string): boolean {
  if (!id.endsWith(".md")) {
    return false;
  }
  return id.split("/").every((segment) => !REFUSED.includes(segment.toLowerCase()));
}

/** A colon before the first slash is a scheme, which a relative path has not. */
function hasScheme(href: string): boolean {
  for (const [index, character] of [...href].entries()) {
    if (character === ":") {
      return index > 0;
    }
    if (!/[A-Za-z0-9+\-.]/.test(character) || (index === 0 && !/[A-Za-z]/.test(character))) {
      return false;
    }
  }
  return false;
}
